### PR TITLE
Optimize queries in habits and merge status into assign

### PR DESCRIPTION
**Pull request by SashkoMolodec**

### CODE REVIEW

**Review by Markiyan000 — APPROVED** [review]

**Review by IvanBegar — APPROVED** [review]

**Review by forestik — APPROVED** [review]

File: service/src/main/java/greencity/service/HabitAssignServiceImpl.java [hunk 1]
```diff
@@ -232,9 +237,143 @@ public void deleteAllHabitAssignsByHabit(HabitVO habit) {
...
+            intervalBetweenDates = Period.between(lastEnrollmentDate, todayDate).getDays();
+        }
+        if ((intervalBetweenDates == 1) || lastEnrollmentDate == null) {
+            habitAssign.setHabitStreak(++habitStreak);
+        } else if (intervalBetweenDates > 1) {
+            habitStreak = 1;
+            habitAssign.setHabitStreak(habitStreak);
+        } else {
+            throw new BadRequestException(ErrorMessage.HABIT_HAS_BEEN_ALREADY_ENROLLED);
+        }
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/HabitAssignServiceImpl.java:291]: Move to private method

File: service/src/main/java/greencity/service/HabitAssignServiceImpl.java [hunk 2]
```diff
@@ -232,9 +237,143 @@ public void deleteAllHabitAssignsByHabit(HabitVO habit) {
...
+
+        int habitDuration = habitAssign.getDuration();
+        if (workingDays == habitDuration) {
+            if (Boolean.TRUE.equals(habitAssign.getAcquired())) {
+                throw new BadRequestException(
+                    ErrorMessage.HABIT_ALREADY_ACQUIRED + habitAssign.getHabit().getId());
+            }
+            habitAssign.setAcquired(true);
+            habitAssignRepo.save(habitAssign);
+        }
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/HabitAssignServiceImpl.java:301]: Same here

File: service/src/main/java/greencity/service/HabitAssignServiceImpl.java [hunk 3]
```diff
@@ -232,9 +237,143 @@ public void deleteAllHabitAssignsByHabit(HabitVO habit) {
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void unenrollHabit(Long habitId, Long userId, LocalDate date) {
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/HabitAssignServiceImpl.java:308]: Check also here what could be moved to private methods

File: service/src/main/java/greencity/service/HabitAssignServiceImpl.java [hunk 4]
```diff
@@ -232,9 +237,143 @@ public void deleteAllHabitAssignsByHabit(HabitVO habit) {
...
+        return daysStreak;
+    }
+
+    private int checkHabitStreakBeforeDate(LocalDate dateTime, HabitAssign habitAssign) {
+        int daysStreak = 0;
+
+        List<LocalDate> enrollDates = habitStatusCalendarService.findEnrolledDatesBefore(dateTime,
+            modelMapper.map(habitAssign, HabitAssignVO.class));
+        Collections.sort(enrollDates);
+        Collections.reverse(enrollDates);
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/HabitAssignServiceImpl.java:358]: Collections.sort(enrollDates, Collections.reverseOrder()); ?

File: service/src/main/java/greencity/service/HabitAssignServiceImpl.java [hunk 5]
```diff
@@ -232,9 +237,143 @@ public void deleteAllHabitAssignsByHabit(HabitVO habit) {
...
+
+        for (int i = 0; i < enrollDates.size() - 1; i++) {
+            if (Period.between(enrollDates.get(i + 1), enrollDates.get(i)).getDays() == 1) {
+                daysStreak++;
+            } else {
+                return daysStreak;
+            }
+        }
```
Comment (vterlyha) [inline service/src/main/java/greencity/service/HabitAssignServiceImpl.java:366]: Also candidate for private method

**Review by vterlyha — APPROVED** [review]

**Pull request merged**